### ISSUE TITLE
tf.debugging.assert_shapes inconsistent behaviour for scalars

### ISSUE THREAD

**Issue by joelberkeley** [tensorflow/tensorflow#44888]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): pip install tensorflow
- TensorFlow version (use command below): 2.3.1
- Python version: 3.7
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Describe the current behavior**
`tf.debugging.assert_shapes` has seemingly inconsistent behaviour for scalars
```
tf.debugging.assert_shapes([(0, ())])  # passes as expected
tf.debugging.assert_shapes([(0, (1,))])  # passes unexpectedly
```
I think this is inconsistent, because `tf.constant(0)` does not have shape `(1,)`.

**Describe the expected behavior**
I would expect `tf.debugging.assert_shapes` to error if the `.shape` attribute is different from the shape constraint

**Comment by joelberkeley** [comment]
I've just noticed this in the docs

> Scalar tensors and specified shapes of length zero (excluding the 'inner-most' prefix) are both treated as having a single dimension of size one.

which tells me this is intentional. I still think it's a problem, because it's inconsistent

**Comment by Saduf2019** [comment]
@joelberkeley 
please share complete standalone code such that we can replicate the issue faced or if possible share a colab gist with the issue reported.

**Comment by joelberkeley** [comment]
aside from `import tensorflow as tf`, the example is standalone

**Comment by joelberkeley** [comment]
i've added the import statement

**Comment by joelberkeley** [comment]
yes, there isn't meant to be any output. That's my concern. I think it would make more sense if the second line raised an error

**Comment by sachinprasadhs** [comment]
The issue still exists in Tf Nightly 2.6, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/02ccd12a258474ca84b78c34bdf60440/44888.ipynb). Thanks!

**Comment by SuryanarayanaY** [comment]
This is still an issue with `tf-nightly(2.16.0-dev20231203)`. Attaching [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/4590a6e6188f3d3f70ee8ec8d8e654dd/44888-tf-nightly-2-16-dev.ipynb) for reference.

**Comment by joelberkeley** [comment]
Indeed, it's documented and intended, though inconsistent. I'd understand if this is something Google won't change because it's a breaking change and being able to assert a shape is isomorphic/equivalent to `[]`, rather than just being equal, can be useful, even if it's not what's desired in all cases

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by joelberkeley** [comment]
I don't have any interest in continuing with this issue. Indeed it's been open for four years. Do with it what you will.

**Comment by Kayyuri** [comment]
Hi @joelberkeley,

Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
